[2단계 - 주문 기능 구현] 에코(조은찬) 미션 제출합니다.

echo724 opened the pull request:
안녕하세요 시카! 에코입니다~
이번 2단계 미션도 1단계와 마찬가지로 페어로 진행했는데요, 3인 페어이다보니 시간 맞추느라, 또 회의하느라 시간이 오래 걸렸던 것 같습니다.
개인적으로는 이번 미션이 가장 어려웠던 것 같은데요, 재화 요구사항 중 할인을 선택했는데, 할인이라는게 저희 내에서 도메인 정의가 확실하지 않아서 그랬던 것 같습니다. 어려웠던 만큼 놓친 기능과 예외 처리 부분 등이 있었던 것 같습니다 ,, ㅜ

저희는 '할인'을 선택했고, 저희가 생각한 할인은 다음과 같습니다.
1. 할인에는 여러 종류가 있을 수 있다.(기본할인, 맴버십 할인 등)
2. 현재는 기본 할인만 지원한다.
3. 기본 할인은 기준 금액 이상일때 특정 퍼센트로 할인하는 것이다. ex) 5만원 이상 구매시 10% 할인

그리고 아래는 미션 진행하면서 서로 애매하거나 잘 몰라서 궁금했던 내용들입니다.
# 데이터베이스에 도메인의 모든 정보를 담는게 좋을지, 아니면 계산 할 수 있는 값들은 저장하지 않는게 좋을지 궁금합니다.
정책이 반영된 가격을 데이터베이스에서 저장할지 말지에 대한 의견이 엇갈렸었습니다.
- 한 크루는 계산할 수 있고 실제 도메인에서도 기본 정보로 계산을 해서 보여주는데 따로 저장할 필요가 없다는 주장이였고,
- 다른 크루는 현업에서는 데이터 분석이나 기록 등의 이유로 가능한 많은 정보를 db에 담으려고 한다. 그러니 저장을 하는게 좋다.

위와 같은 이유들이였고, 저희는 아래 방식으로 구현했습니다.
    - 현재는 데이터베이스에서 적용됐었던 정책을 불러와 다시 계산해서 보여줌.

그래서 현업에서는 db에 저장할 데이터, 특히 도메인 필드 데이터를 어디까지 저장하는지 궁금합니다!

# 전역 예외처리를 할 때 일관된 예외 메세지를 보여주려고 하는데, 그럴 경우에는 어떻게 해야할지 궁금합니다!
프론트 크루들과 예외에 대한 API를 작성할때, 특정 필드로 예외 메세지를 넣어서 보내주기로 했습니다.
이럴 경우 커스텀 예외 외에도 기본 REST 예외 들을 매핑해주려고 하는데 여러 방법을 시도했는데 확실히 뭐가 맞는 방법이다 하는게 없었던 것 같습니다.
예를 들어, 설정하지 않은 URL로 접근시 발생하는 404 NotFound의 경우 저희의 CustomException으로 어떻게 맵핑해주려고 합니다.
저희는 처음에 ResponseEntityExceptionHandler의 handleExceptionInternal 메서드를 재정의하여 해결하려고 했지만, 호출되지 않았습니다. 두번째는 @EnableWebMVC 어노테이션을 추가하고 properties에 spring.mvc.throw-exception-if-no-handler-found=true를 추가해줬습니다. 다만 이 경우 admin 페이지에서 css 파일을 불러오지 못하는 에러가 발생하더군요.. [블로그](https://velog.io/@shson/spring-rest-api%EC%97%90%EC%84%9C-not-found-exception%EC%B2%98%EB%A6%AC%ED%95%98%EA%B8%B0)를 참조했습니다.


시카는 이런 경우에 어떻게 문제를 해결하실 것인지, 혹은 어떤 키워드로 학습하면 알 수 있는지 궁금합니다..

구현 부분에서 궁금하신 부분 있다면 언제든지 dm 주세요! 감사합니다~

kyY00n committed:
feat: CORS 설정 추가

kyY00n committed:
feat: sql 로깅

* db를 파일로 저장

kyY00n committed:
feat: 데이터베이스 콘솔 활성화

* data.sql 삭제

kyY00n committed:
feat: webAllowOthers 을 활성화

kyY00n committed:
fix: 모든 메서드 허용

kyY00n committed:
feat: Location 헤더를 cors 허용

kyY00n committed:
feat: 모든 헤더 허용

kyY00n committed:
feat: expoesed header 설정

Commit:
feat: Spring Actuator 의존 추가 및 Endpoint 설정

actuator/health -> 서버 건강 상태 확인
actuator/logfile -> 서버 로그 확인

kyY00n committed:
Merge pull request #2 from kyY00n/feat/actuator

Feat/actuator 기능 구현

kyY00n committed:
feat: 테스트 데이터베이스 분리

kyY00n committed:
feat: restdocs 적용

kyY00n committed:
docs: 기존 API 문서 추가

Commit:
feat: HTTP 요청/응답 로깅 처리

amaran-th committed:
Merge pull request #4 from kyY00n/amaranth

feat: HTTP 요청/응답 로깅 처리

kyY00n committed:
test: 결제 금액 조회 인수 테스트 추가

kyY00n committed:
test: 주문하기 및 주문 목록 조회 인수 테스트 추가

kyY00n committed:
test: 주문하기 및 주문 목록 조회 인수 테스트 추가

kyY00n committed:
Merge remote-tracking branch 'origin/step2' into step2

# Conflicts:
#	src/docs/asciidoc/index.adoc
#	src/main/java/cart/dto/OrderDetailResponse.java
#	src/test/java/cart/integration/OrderIntegrationTest.java

kyY00n committed:
docs: api 명세 업데이트

kyY00n committed:
feat: money 객체 정의

kyY00n committed:
feat: Product 의 price를 Money 타입으로 수정

kyY00n committed:
docs: 2단계 기능 목록 정의

kyY00n committed:
feat: Money 연산 기능 추가

kyY00n committed:
feat: Order, OrderItem 도메인 객체 정의

kyY00n committed:
feat: DeliveryPolicy 정의 및 구현

kyY00n committed:
feat: DiscountPolicy 정의 및 구현

kyY00n committed:
feat: Payment 정의 및 구현

kyY00n committed:
feat: 결제 예상 금액 조회 API 구현

kyY00n committed:
feat: 장바구니 상품 아이디 리스트가 비어있을 경우 예외 처리

kyY00n committed:
feat: 장바구니에 담은 상품을 주문 기능 추가

kyY00n committed:
test: OrderWithId Fixture 추가

kyY00n committed:
feat: 주문 목록 및 상세 내역 조회 기능 추가

kyY00n committed:
feat: 장바구니 상품 유저 권한 확인 기능 추가

kyY00n committed:
feat: 예외 정의 및 사용

kyY00n committed:
feat: 추가된 커스텀 예외들 전역 처리 추가

Commit:
feat: FakeRepository를 DbRepository로 변경

Commit:
feat: Delivery, Discount Dao 작성

Commit:
hotfix: 테스트 통과하도록 수정

Commit:
fix: paymentService DB에 저장하도록 수정

Commit:
chore: 전역 REST API 예외 처리 추가

Commit:
hotfix: 스키마 테이블 이름 변경

Commit:
fix: Member 인증 예외 처리 구현

Commit:
fix: 적용되는 할인이 없을 경우 결제 기록이 저장되지 않는 문제 해결

Commit:
feat: Mysql을 사용하는 Production 프로파일 설정

dks301 commented on README.md:
```diff
@@ -1 +1,109 @@
+    - [x] 비어있을 경우 예외처리
+    - [x] 주문을 완료한 장바구니 상품은 장바구니에서 삭제된다.
+    - [x] 요청한 회원의 상품이 아닐 경우 예외처리한다.
+- [ ] 할인 정책을 적용할 수 있다.
+    - [x] 할인 정책은 기준 금액과 할인율을 가지고 있다.
+        - [ ] 할인율은 100% 이하여야한다.
+    - [x] 기준 금액 이상 주문 시 전체 금액에서 할인율만큼 할인된다.
+    - [x] 현재 주문에 적용 가능한 할인 정책을 찾을 수 있다.
+      ex) 5만원 이상 주문 시 전체 금액에서 10% 할인이 된다.
+- [x] 사용자 별로 주문 목록을 확인할 수 있다.
+    - [ ] 최신순으로 주문 목록을 확인할 수 있다.
```
완료되지 않은 요구사항이 있는데요...!
협업을 할 때 논의된 요구사항을 충족하는 부분은 생각보다 중요합니다 ㅎㅎㅎ 완료된걸까요?

echo724 replied:
앗 빼먹었네요 ㅜ 언급해주셔서 감사합니다. 제출 전에 꼭 요구사항을 다시 읽어봐야겠습니다.
해당 부분 구현 완료하였습니다!

dks301 commented on src/main/java/cart/application/OrderService.java:
```diff
@@ -0,0 +1,65 @@
+
+    private final OrderRepository orderRepository;
+
+    private final CartItemRepository cartItemRepository;
```
CartItem 관련 로직이 들어가서 CartItemRepository를 직접 DI하신 걸로 보이는데요...!
CartItemService로 간접적으로 DI해면 중복코드나 공통처리를 할 수 있는 장점이 있을 것 같은데,
Repository를 DI할 때 어떤 장점이 있어서 사용하신 걸까요??

echo724 replied:
사실 이 부분에서도 의견이 서로 달랐는데, 저희가 낸 결론은 Service에서 하위 계층의 Repository를 조합해서 사용하는게 자연스럽다는 결론을 내렸던 것 같습니다. Controller에서도 마찬가지로 여러 Service들을 조합해서 기능을 달성하고 있어서 이 구조의 일관성을 유지하였습니다. 만약 다른 Service에게 요청을 한다면 필드로 Service와 Repository를 동시에 가지고 있을텐데 이게 자연스러워보이지 않더라구요. 다만 물론 계층간 분리가 잘 되지 않아서 수정에 취약해질 것 같다는 문제점이 있긴 합니다! 그래서 정답은 아직 모르지만 다른 서비스를 사용하는게 시카가 얘기해주신대로 중복 코드와 공통처리가 되어 확실히 장점이 있을 것 같습니다!

dks301 replied:
에코가 이야기해준 것 처럼 사실 대부분은 정답이 있는 것들이 아니기에 합의하에 정한 팀 컨벤션을 지키고 일관성을 유지하는게 더더욱 중요한 것 같아요 👍ㅎㅎㅎ

다만, 미션으로 공부중인만큼 자연스럽다보다는, 구체적으로 다른 개발자들을 어떻게 설득하면 좋을지 고민해보면 어떨까요?? ㅎㅎ
예를들어, Repository로 db 관련 로직 전체가(db connection, query, 기타 등등) 고수준으로 추상화되어 있으니 DB에서 가져다 쓰기보단 Java Collection을 사용하듯 편하게 domain을 가져다 사용할 수 있다 처럼요!

echo724 replied:
헉! 🥹 시카.. 예시 넘 감사합니다. 자연스럽다라는 표현을 쓰면서도 더 구체적인 방식으로 설명을 드리고 싶은데 생각이 나지 않았거든요! 시카처럼 논리적인 설득방식을 좀 더 활용해서 대답하도록 공부해보겠습니다! 💪

dks301 commented on src/main/java/cart/application/OrderService.java:
```diff
@@ -0,0 +1,65 @@
+                .map(cartItemId -> this.findCartItemOf(cartItemId, member))
+                .map(OrderItem::from)
+                .collect(Collectors.toList());
+        return new Order(member, orderItems);
```
1. Order를 생성만 하고 저장하지 않고 반환하는 이유가 있을까요??
2. db에 저장하지 않기떄문에 Transactional을 관리하지 않는 것으로 보이는데요 ㅎㅎ 동시성 이슈는 없을까요??

echo724 replied:
현재 Order의 구조는 다음과 같습니다.

가격 조회
가격 조회 요청 → CartItem을 OrderItem으로 변경(CartItem과 의존성 제거) → DraftOrder 생성 → 가격 정보 생성 및 저장

주문 요청 → CartItem을 OrderItem으로 변경(CartItem과 의존성 제거) → DraftOrder 생성 → Order 저장 → 가격 정보 생성 및 저장 → CartItem 삭제

이렇게 구조를 짠 이유는 
DraftOrder 생성 과정의 중복을 줄이기 위해서였습니다! 그래서 DraftOrder는 따로 저장하지 않고 실제 Order 객체를 저장해주는데요 그런 이유로 DraftOrder 생성과정에서는 Transactional을 따로 고려하지 않았습니다.

다만 동시성 문제로는 두가지 경우가 있을 것 같은데요,  다음과 같이 처리하고 있습니다.
1. CartItemId를 OrderItem으로 반환하는 중 CartItem이 이미 삭제 되어 반환이 안되는 경우 -> Order 생성 중 NotFound 예외 생성하여 Order 생성 방지
2. Order를 생성하고 CartItem을 지우는 중 CartItem이 이미 지워진 경우 -> Order 생성 메서드 안에서 예외를 생성하여 Transaction으로 Order 롤백

dks301 replied:
고민해주신 것처럼 비즈니스 로직 레벨에서의 동시성 이슈 외에도 고려해야하는 상황이 있는데요! 조금 더 고민해볼까요??

1. 거의 완전(?) 동시성 이슈 - ex. CartItemId를 OrderItem으로 최종 반환하는 중(HttpResponse 생성과 같은) CartItem이 조회 성공해서 예외처리가 안되는 경우
2. 여러대의 애플리케이션 서버로 서비스를 하는 경우

echo724 replied:
시간을 좀 들여서 고민을 해봤는데 CartItem과 관련해서는 동시성 문제가 거의 없을 것 같습니다. 이유는 아래와 같은데요
- CartItem을 OrderItem으로 복사하는 과정은 Order를 생성하기 전 제일 처음 실행되는 작업입니다. 따라서 CartItem이 순간 삭제 되기 전에 OrderItem으로 복사가 되더라도 결국 마지막에 CartItem을 지우는 과정에서 CartItem이 존재하지 않아 예외가 발생하기 때문에 롤백되어 주문이 되지 않는 구조입니다.
- 다만, Product로 부터 가져온 가격 정보가 OrderItem을 생성한 뒤 바뀌는 경우 최종 결제 가격 정보가 바뀌어 결제되는 경우가 있을 것 같습니다! 이런 경우를 대비해 주문 마지막 단계에서 현재 카트로부터 계산한 가격 정보와 비교해서 가격이 다르면 예외를 발생시킬 것 같네요!

echo724 replied:
가격 변동의 경우 주문의 아이템 가격과 주문 생성 후 CartItem에서 Product의 가격을 가져와 계산한 가격과 비교하여 다르면 예외를 발생시키도록 처리하였습니다!

dks301 commented on src/main/java/cart/application/OrderService.java:
```diff
@@ -0,0 +1,65 @@
+
+    public Order createDraftOrder(final Member member, final List<Long> cartItemIds) {
+        if (cartItemIds.isEmpty()) {
+            throw new EmptyItemInput();
```
```suggestion
            throw new EmptyItemInputException();
```
네이밍에 Exception이 없어서 예외인지 한 번 더 확인했네요 ㅎㅎㅎ😅 명시해주지 않은 특별한 이유가 있을까요??

echo724 replied:
앗 이건 OrderException.EmptyItemInput 예외인데, 앞 부분을 실수로 static import하여 명시적으로 Exception이 드러나지 않았던 것 같습니다. OrderException.EmptyItemInput()으로 변경하였습니다!

dks301 replied:
static import는 제가 놓쳤네요 😅
결과적으로 지금과 같이 OrderException과 같이 static import 등으로 잘못 사용하는 경우가 있을 수 있는 경우에는 다소 중복처럼 보이더라도 Exception을 명시적으로 선언해주면 어떨까요??

`명시적 표현 vs 간결한 표현`이 고민되는 경우, 명시적으로 적어주는게 휴먼에러나 오용을 방지하는 등 장점이 많지 않을까요??

echo724 replied:
맞습니다. 이번 경우처럼 static import로 예외인지 모르게 하는 실수였어서 명시적 표현이 좋더라구요! 전부 변경하였습니다. 마찬가지로 제 코드를 보시면 내부 메서드나 필드는 전부 this를 표기하도록 되어있는데요, 이또한 명시적으로 표현하기 위한 방법이였습니다~!

dks301 commented on src/main/java/cart/application/PaymentService.java:
```diff
@@ -0,0 +1,43 @@
+        this.paymentRepository = paymentRepository;
+    }
+
+    public PaymentRecord createDraftPaymentRecord(final Order order) {
+        final List<DiscountPolicy> discountPolicies = this.discountPolicyRepository.findDefault();
+        final List<DeliveryPolicy> deliveryPolicies = this.deliveryPolicyRepository.findDefault();
+        final Payment payment = new Payment(discountPolicies, deliveryPolicies);
+        return payment.createPaymentRecord(order);
+    }
+
+    public PaymentRecord createPaymentRecordAndSave(final Order order) {
```
여기도 그렇고 전반적으로  `@Transactional`이 없는 Service의 메서드들이 있는데요...! 
어떤 기준으로 어노테이션 유무를 정하신걸까요??

echo724 replied:
이 부분도 놓친 것 같네요 ㅜ Service 클래스에 전역적으로 @Transactional을 붙이고 조회 메서드에는 ReadOnly=true로 변경하였습니다!

dks301 commented on src/main/java/cart/domain/DefaultDiscountPolicy.java:
```diff
@@ -0,0 +1,59 @@
+    private final Long id;
+    private final String name;
+    private final Money threshold;
+    private final double discountRate;
```
할인 ㅎㅎㅎ 저도 실무에서 돈관련 개발을 하고 있어서 반갑네요 😄

할인 같은 경우 소숫점 자릿수에 따라 금액이 차이가 날 수 있습니다. 따라서 정확한 계산이 중요하고 1원이라도 문제가 생기면 안되는데요...!
관련 문제를 어떻게 해결할 수 있는지 고민해보면 어떨까요??

echo724 replied:
앗 시카 멋있습니다@.@
짜면서도 나중에 문제가 생기지 않을까 생각은 했습니다. 그래서 시카 리뷰를 보고 BigDecimal을 사용하였습니다! 기존 double은 부동 소수점 사용으로 문제가 생길 수 있겠더라구요! BigDecimal은 부동 소수 대신 십진수 표현으로 부동소수점 연산에서 올 수 있는 문제를 해결하고 있다고 알고 있는데요@ 정확하게 실무에서는 이 자료형을 쓰는지는 모르겠습니다 ㅎㅎ ㅜ  실무에서는 어떤 자료형을 사용하고 어떻게 정확성을 유지하나요?

dks301 replied:
일단 저희 팀은 BigDecimal 사용하고 있어요 ㅎㅎㅎ
또 정책마다 다르지만 연산시 MathContext나 기타 여러가지 추가 기능, 소숫점 이하 정책(반올림, 올림, 버림 등등) 을 사용해서 정확성 문제를 해결하고 있습니다 ㅎㅎ

전부다 공부해볼 수는 없으니 대표적인
BigDecimal 나눗셈 처리할 떄 고려해야할 점과 이슈 해결 방법에 대해서 공부해보면 어떨까요??

echo724 replied:
그렇군요! ㅎㅎ 답변 감사합니다~~

BigDecimal에 대해서 공부를 좀 해봤는데요 다음과 같은 이슈들이 있을것 같더라구요!
1. 0으로 나눌 경우
2. scale이 정해지지 않아 발생하는 ArithmeticException
정도 인 것 같습니다!

해결법으로는
1. 나누는 값이 0이 아닌지 확인해주고
2. divide(divisor, scale, roundingMode) 메서드를 사용해서 소수점 자리와 올림, 반올림 모드를 사용해 나누는 것입니다.
추가로 divide(divisor, mathContext)도 있는 것 같더라구요. mathContext는 Precision과 RoundingMode를 객체로 만들어둔 것이서 이걸 활용해 전역적으로 설정가능할 것 같습니다.

dks301 commented on src/main/java/cart/domain/Money.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.Objects;
+
+public class Money {
+    private final int value;
```
돈과 관련된 부분에서 int라는 데이터타입은 너무 작지 않을까요??
실제로도 기업의 매출이나 영업이익 같은 경우 int의 최대값인 21억을 훨씬 넘는 경우가 많고, 원화가 아닌 글로벌 서비스라면 더 큰단위의 수가 필요한 경우가 있는데요...! 적절한 데이터 타입을 고민해볼까요??

echo724 replied:
앗 이 부분도 BigInteger 사용으로 int 최대값보다 큰 값을 지원하려고 했습니다만,, Money 객체의 경우 워낙 많은 곳에서 사용되고 돈과 관련된 부분을 전부 BigInteger로 바꿔야 의미가 있다는 생각이 들어 리펙토링을 하진 않았습니다. 그래서 생각해보기로는 여기서 Money는 오로지 상품 가격과 관련된 돈 관련 값만 관리하고 있으므로 21억 정도의 가격을 가진 상품을 사고팔지 않기에 int를 넘어가는 값을 다루지는 않을 것으로 판단했습니다. 물론 10억 단위 가격으로 거래되는 상품이 있을 경우 리펙토링을 진행해볼 것 같습니다!

dks301 replied:
말씀해주신 것처럼 타당한 이유가 있다면 현재로 유지하는 것도 좋은 방법이라고 생각해요 👍

여기서 추가로 고민해보면 좋을 것 같은 부분이 있는데요...!
1. 현재 Money 객체가 VO로서 서비스의 돈 관련 부분에서 모두 사용하고 있는데요 ㅎㅎㅎ 모든 부분에서 해당 class를 사용하는게 맞는지 고민해보면 어떨까요?? 가령 OrderPrice.class, ProductPrice.class 처럼요
2. 래퍼 클래스(Integer) vs 원시값(int) 중 int를 사용한 이유와 각각의 장단점

echo724 replied:
1. 시카 말씀대로 이번 리펙토링이 힘들었던 이유가 너무 많은 곳에서 Money가 사용되고 있기 때문이라고 생각합니다. 만약 Money객체가 거의 변경 가능성이 없다면 VO로써 여러 군데에서 사용해도 되지만, 변경 가능성이 조금이라도 있으면 각 객체에서 의존을 끊어주는 것도 괜찮겠더라구요. 그리고 추가로 OrderPrice와 ProductPrice에 관해 각 래퍼 클래스가 검증해야할 값이 다르다면 따로 래퍼 클래스를 두는 것도 좋을 것 같습니다!
2. int의 경우 Integer와 달리 객체 생성 비용이 따로 없고 값을 사용할때 따로 unboxing해줄 필요가 없기 때문에 빠른 장점이 있습니다! 단점으로는 generic과 함께 사용할 수 없다 입니다. 그리고 또다른 차이는 null 값을 가질 수 없다 정도인 것 같습니다. 이번 Money의 경우 1. 값이 null로 저장될 이유가 없고, 2. generic과 함께 사용되지 않기 때문에 int로 사용했습니다.
다만,, 지금 든 생각인데, Money의 필드를 int가 아니라 generic T로 했으면 BigInteger로도,Integer로도 사용가능했을 것 같네요! 값객체의 경우 특히나 내부 값의 타입을 다르게 가질 수 있도록 하는게 좋아보이네요!!! 오오.. 혹시 이걸 의도하시고 래퍼 클래스에 대해 고민해보라고 하신 건가요?! ㅎㅎ 좋은 깨달음 주셔서 감사합니다~!

dks301 replied:
> 오오.. 혹시 이걸 의도하시고 래퍼 클래스에 대해 고민해보라고 하신 건가요?! ㅎㅎ

앗 그 정도까지는 아니에요 생각의 방향만 질문드렸을뿐 에코가 많이 고민하시고 발전해나가고 있네요! 💯💯

dks301 reviewed (changes requested):
안녕하세요 에코, 리뷰어 시카입니다. 😄

어려운 미션인데 요구사항에 맞춰 잘 구현해주셨네요! 👍
몇 가지 코멘트 남겨드렸으니 확인부탁드려요.

궁금한 점이 생기면 언제든지 DM 주세요 ㅎㅎ

---

> 데이터베이스에 도메인의 모든 정보를 담는게 좋을지, 아니면 계산 할 수 있는 값들은 저장하지 않는게 좋을지 궁금합니다.
> 정책이 반영된 가격을 데이터베이스에서 저장할지 말지에 대한 의견이 엇갈렸었습니다.
> 
> 한 크루는 계산할 수 있고 실제 도메인에서도 기본 정보로 계산을 해서 보여주는데 따로 저장할 필요가 없다는 주장이였고,
> 다른 크루는 현업에서는 데이터 분석이나 기록 등의 이유로 가능한 많은 정보를 db에 담으려고 한다. 그러니 저장을 하는게 좋다.
> 위와 같은 이유들이였고, 저희는 아래 방식으로 구현했습니다.
> - 현재는 데이터베이스에서 적용됐었던 정책을 불러와 다시 계산해서 보여줌.
> 
> 그래서 현업에서는 db에 저장할 데이터, 특히 도메인 필드 데이터를 어디까지 저장하는지 궁금합니다!

이 부분은 양쪽의 의견 모두 타당하고 실무에서도 적용되는 내용인데요!!
1. 객체지향 원칙처럼 db또한 계산할 수 있는 데이터는 기본적으로 저장하지 않는게 원칙인데요...! 아시겠지만 정규화라고 합니다 ㅎㅎㅎ
2. 하지만, 데이터 분석이나, 기록 등 여러가지 이유로 정규화를 하지 않고 이를 비정규화해서 저장하는 경우도 물론 있습니다 ㅎㅎㅎ

결론적으로 기본 원칙은 정규화를 한다. 하지만 필요한 경우 or  어쩔 수 없는 경우 비정규화하는 것이 일반적인데요...!
지금과 같은 경우는 개인적으로 필요한 경우 or 어쩔 수 없는 경우가 아니라고 생각합니다 ㅎㅎ
페어분들과 이 리뷰와 다른 리뷰이분들의 이야기를 종합해서 결론을 알려주시면 좋을 것 같아요 😄


> 전역 예외처리를 할 때 일관된 예외 메세지를 보여주려고 하는데, 그럴 경우에는 어떻게 해야할지 궁금합니다!
> 프론트 크루들과 예외에 대한 API를 작성할때, 특정 필드로 예외 메세지를 넣어서 보내주기로 했습니다.
> 이럴 경우 커스텀 예외 외에도 기본 REST 예외 들을 매핑해주려고 하는데 여러 방법을 시도했는데 확실히 뭐가 맞는 방법이다 하는게 없었던 것 같습니다.
> 예를 들어, 설정하지 않은 URL로 접근시 발생하는 404 NotFound의 경우 저희의 CustomException으로 어떻게 맵핑해주려고 합니다.
> 저희는 처음에 ResponseEntityExceptionHandler의 handleExceptionInternal 메서드를 재정의하여 해결하려고 했지만, 호출되지 않았습니다. 두번째는 @EnableWebMVC 어노테이션을 추가하고 properties에 spring.mvc.throw-exception-if-no-handler-found=true를 추가해줬습니다. 다만 이 경우 admin 페이지에서 css 파일을 불러오지 못하는 에러가 발생하더군요.. [블로그](https://velog.io/@shson/spring-rest-api%EC%97%90%EC%84%9C-not-found-exception%EC%B2%98%EB%A6%AC%ED%95%98%EA%B8%B0)를 참조했습니다.
> 
> 시카는 이런 경우에 어떻게 문제를 해결하실 것인지, 혹은 어떤 키워드로 학습하면 알 수 있는지 궁금합니다..

음... 이 부분은 딱히 정형화된 해결법이 없는 것 같은데요 ㅎㅎㅎ 논의중 혹은 다른 리뷰에서 정형화된 방법이 있다면 저도 알려주시면 감사하겠습니다 😄
일단 저같은 경우 당장 생각나는 방법은 
1. 설정하지 않은 url 접근을 CustomException으로 맵핑해야하는지 고민해볼 것 같은데요...! 서버단에서 정상적으로 알고있는(처리해야하는) 예외와 아닌 예외는 오히려 구분하는게 좋지 않을까요??
2. 그럼에도 구분한다면 전처리를 해야하는 부분이니 nginx와 같은 프록시 서버를 사용해서 전처리할 것 같아요 ㅎㅎㅎ

Commit:
feat: 주문 내역 주문 시간순으로 정렬 기능 추가

Commit:
feat: 정책 설정 관리자 페이지 추가

Commit:
refactor: discountRate 자료형 변경(double -> BigDecimal)

Commit:
fix: 올바르지 않은 DiscountRate 예외 처리

Commit:
refactor: Transactional 추가

Commit:
fix: Order 생성 동시성 문제 해결

Order를 만드는 중 발생할 수 있는 동시성 문제
1. CartItemId를 OrderItem으로 반환하는 중 CartItemId가 이미 삭제 되어 반환이 안되는 경우 -> Order 생성 중 NotFound 예외 생성
2. Order를 생성하고 CartItem을 지우는 중 CartItem이 이미 지워진 경우 -> Order 생성 메서드 안에서 예외 생성(Transaction)

echo724 commented:
안녕하세요 시카! 궁금증이 풀리는 자세한 답변 달아주셔서 감사합니다!
덕분에 정규화 키워드와 예외 처리 매핑에 대해 다시 생각해본 것 같습니다. 페어들과 이부분 대해서 더 얘기 나눠보려고 합니다 ㅎㅎ
그리고 시카가 달아주신 리뷰 반영하여 구현하였습니다~! 감사합니다 시카~!

dks301 reviewed (changes requested):
안녕하세요 에코, 리뷰어 시카입니다. 😄

코멘트 드린 내용을 잘 적용해주셨네요 👏👏
조금 더 고민해볼만한 몇 가지 코멘트와 질문답변을 남겨드렸으니 확인해주세요.

추가로 궁금한 점이 생기면 언제든지 DM 주세요 ㅎㅎ

Commit:
feat: 주문 중 상품 가격 변동이 있을 경우 예외 처리

Commit:
refactor: 커스텀 예외에 명시적으로 Exception 표기

Commit:
refactor: PriceMismatch 예외시 400 코드 반환하도록 처리

echo724 commented:
안녕하세요 시카~ 시카가 남겨주신 리뷰 반영해서 코드 수정하고 고민거리 공부해봤습니다!
알려주신 키워드 덕분에 꼭 알고 가야할 부분들 더 공부할 수 있었네요! 감사합니다~~!

dks301 reviewed (approved):
안녕하세요 에코, 리뷰어 시카입니다. 😄

어려운 미션이었는데 피드백까지 의도한 것 이상으로 잘 반영해주셨네요! 👍
이번 미션은 여기서 머지하겠습니다 ㅎㅎ
레벨2 고생 많으셨습니다 🎉🎉

추가로 궁금한 점이 생기면 언제든지 DM 주세요 ㅎㅎ
